sdk/assemblyscript/src/assembly/__tests__/openai.spec.ts

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2025 Hypermode Inc.
+ * Licensed under the terms of the Apache License, Version 2.0
+ * See the LICENSE file that accompanied this code for further details.
+ *
+ * SPDX-FileCopyrightText: 2025 Hypermode Inc. <hello@hypermode.com>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { expect, it, log, run } from "as-test";
+import { JSON } from "json-as";
+
+import {
+  SystemMessage,
+  UserMessage,
+  AssistantMessage,
+  RequestMessage,
+  parseMessages,
+} from "../../models/openai/chat";
+
+it("should round-trip chat messages", () => {
+  const msgs: RequestMessage[] = [
+    new SystemMessage("You are a helpful assistant."),
+    new UserMessage("What is the capital of France?"),
+    new AssistantMessage("The capital of France is Paris."),
+  ];
+
+  const data = JSON.stringify(msgs);
+  const parsedMsgs = parseMessages(data);
+
+  expect(parsedMsgs.length).toBe(msgs.length);
+  for (let i = 0; i < msgs.length; i++) {
+    expect(msgs[i].role).toBe(parsedMsgs[i].role);
+  }
+
+  const roundTrip = JSON.stringify(parsedMsgs);
```

Comment:
It's using the `__SERIALIZE` and `__DESERIALIZE` functions on `RequestMessage`. That's why its only returning the `role` field.

The underlying problem is that the return type in `parseMessages` is being demoted from `RawMessage` to `RequestMessage`.

I fixed it in two ways:
1. Change the return type of `parseMessages` to `RawMessage`
2. Use `changetype<RawMessage[]>(parsedMsgs)` to lift the type

Reply:
Changing the return type would not be viable, as the user will need to be able to append objects of other message types to the resulting array.